build.sbt

```diff
@@ -450,52 +451,130 @@ lazy val scalalib =
+
+        // SN Port: sjs uses baseDirectory.value.getParentFile here.
+        val base  = baseDirectory.value
+        val parts = ver.split(Array('.', '-'))
+        val verList = parts.inits.map { ps =>
+          val len = ps.mkString(".").length
+          // re-read version, since we lost '.' and '-'
+          ver.substring(0, len)
+        }
+        def dirStr(v: String) =
+          if (v.isEmpty) "overrides" else s"overrides-$v"
+        val dirs = verList.map(base / dirStr(_)).filter(_.exists)
+        dirs.toSeq // most specific shadow less specific
+      },
+      // Compute sources
+      // Files in earlier src dirs shadow files in later dirs
+      sources in Compile := {
+        // Sources coming from the sources of Scala
+        val scalaSrcDir = fetchScalaSource.value
+
+        // All source directories (overrides shadow scalaSrcDir)
+        val sourceDirectories =
+          (unmanagedSourceDirectories in Compile).value :+ scalaSrcDir
+
+        // Filter sources with overrides
+        def normPath(f: File): String =
+          f.getPath.replace(java.io.File.separator, "/")
+
+        val sources = mutable.ListBuffer.empty[File]
+        val paths   = mutable.Set.empty[String]
+
+        val s = streams.value
+
+        for {
+          srcDir <- sourceDirectories
+          normSrcDir = normPath(srcDir)
+          src <- (srcDir ** "*.scala").get
+        } {
+          val normSrc = normPath(src)
+          val path    = normSrc.substring(normSrcDir.length)
+          val useless =
+            path.contains("/scala/collection/parallel/") ||
+              path.contains("/scala/util/parsing/")
+          if (!useless) {
+            if (paths.add(path))
+              sources += src
+            else
+              s.log.debug(s"not including $src")
+          }
         }
 
-        // Checkout proper ref. We do this anyway so we fail if
-        // something is wrong
-        val git = Git.open(trgDir)
-        s.log.info(s"Checking out Scala source version ${scalaVersion.value}")
-        git.checkout().setName(s"v${scalaVersion.value}").call()
-
-        IO.delete(file("scalalib/src/main/scala"))
-        IO.copyDirectory(trgDir / "src" / "library" / "scala",
-                         file("scalalib/src/main/scala/scala"))
-
-        val epoch :: major :: _ = scalaVersion.value.split("\\.").toList
-        IO.copyDirectory(file(s"scalalib/overrides-$epoch.$major/scala"),
-                         file("scalalib/src/main/scala/scala"),
-                         overwrite = true)
-
-        // Remove all java code, as it's not going to be available
-        // in the NIR anyway. This also resolves issues wrt overrides
-        // of code that was previously in Java but is in Scala now.
-        (file("scalalib/src/main/scala") ** "*.java").get.foreach(IO.delete)
+        sources.result()
       },
-      Compile / compile := (Compile / compile)
-        .dependsOn(assembleScalaLibrary)
-        .value,
```

Comment:
The new code relies on `update` so the `compile` dependency is no longer needed?

Reply:
Exactly!  I believe that removing the section you quoted is actually a simplification and
separation of concerns.  The fetching of scala sources is done in the `scalalib / assemble`
phase, where one innocent of prior art would expect.

The change in nativelib to publish nscplugin in its assemble phase rather than the previous
`publishLocal`  is a more than obscure workaround for the more than obscure way
the nscplugin dependency is introduced in ScalaNativePluginInternal.scala.

I have proposed on gitter a better next-generation solution, based on sbt `onLoad` & a new
Task. Still not a good solution. Perhaps someone will come up with something better.

I think/hope the workaround in this PR will earn its keep while we find something better.
I have tried to make it possible for anyone in the future to follow this trickery but will
apologize not to those future forlorn souls.